packages/aila/src/core/document/AilaDocument.ts

```diff
@@ -1,150 +1,327 @@
+import type { AilaDocumentService } from "../AilaServices";
 import type { Message } from "../chat";
-import type { AilaDocumentContent } from "./types";
+import type {
+  AilaDocumentContent,
+  CategorisationPlugin,
+  DocumentPlugin,
+} from "./types";
 
-const log = aiLogger("aila:lesson");
+const log = aiLogger("aila:document");
 
 export class AilaDocument implements AilaDocumentService {
-  private readonly _aila: AilaServices;
-  private _content: AilaDocumentContent;
+  private _content: AilaDocumentContent = {};
   private _hasInitialisedContentFromMessages = false;
   private readonly _appliedPatches: ValidPatchDocument[] = [];
   private readonly _invalidPatches: ValidPatchDocument[] = [];
-  private readonly _categoriser: AilaCategorisationFeature;
+  private readonly _plugins: DocumentPlugin[] = [];
+  private readonly _categorisationPlugins: CategorisationPlugin[] = [];
+  private readonly _schema?: z.ZodType<AilaDocumentContent>;
 
+  /**
+   * Create a new AilaDocument
+   * @param content Initial content (optional)
+   * @param plugins Document plugins for document-specific operations
+   * @param categorisationPlugins Plugins for content categorisation
+   * @param schema Schema for document validation
+   */
   constructor({
-    aila,
-    content,
-    categoriser,
+    content = {},
+    plugins = [],
+    categorisationPlugins = [],
+    schema,
   }: {
-    aila: AilaServices;
-    content?: LooseLessonPlan;
-    categoriser?: AilaCategorisationFeature;
+    content?: AilaDocumentContent;
+    plugins?: DocumentPlugin[];
+    categorisationPlugins?: CategorisationPlugin[];
+    schema?: z.ZodType<AilaDocumentContent>;
   }) {
-    log.info("Creating AilaDocument");
-    this._aila = aila;
-    this._content = content ?? {};
-    this._categoriser =
-      categoriser ??
-      new AilaCategorisation({
-        aila,
-      });
+    log.info(`Creating ${this.constructor.name}`);
+
+    this._plugins = [];
+    this._categorisationPlugins = [];
+
+    this._content = content;
+    this._schema = schema;
+
+    if (plugins && plugins.length > 0) {
+      plugins.forEach((plugin) => this.registerPlugin(plugin));
+    }
+
+    if (categorisationPlugins && categorisationPlugins.length > 0) {
+      categorisationPlugins.forEach((plugin) =>
+        this.registerCategorisationPlugin(plugin),
+      );
+    }
   }
 
-  public get content(): AilaDocumentContent {
+  /**
+   * Register a document plugin
+   */
+  registerPlugin(plugin: DocumentPlugin): void {
+    this._plugins.push(plugin);
+  }
+
+  /**
+   * Register a categorisation plugin
+   */
+  registerCategorisationPlugin(plugin: CategorisationPlugin): void {
+    log.info("registerCategorisationPlugin called", {
+      pluginId: plugin.id,
+    });
+    this._categorisationPlugins.push(plugin);
+  }
+
+  /**
+   * Get the appropriate plugin for the content
+   * Since we now only register one plugin, we just return the first one
+   */
```

Comment:
If we're only using one plugin, and it only works if there is one plugin, should we just support one plugin?

A bit of an embarrassing question, but what actually is a plugin? I can't find where it's being applied in the calling code